### PR TITLE
[TS-217] Yaga cmd tool

### PR DESCRIPTION
**Pull request by jenchik**

### CODE REVIEW

File: cli/interface.go [hunk 1]
```diff
@@ -23,3 +24,11 @@ type Instance interface {
 	Run(RunOptions) error
 	Shutdown(ctx context.Context) error
 }
+
+type (
+	Context = cli.Context
+
+	Command = cli.Command
+
+	Commandor func(*Options) Command
```
Comment (im-kulikov) [inline cli/interface.go]: funny

File: cli/options.go [hunk 2]
```diff
@@ -115,3 +116,14 @@ func Debug(args ...bool) Option {
 		}
 	}
 }
+
+// Commands closure to set additional commands for CLI
+func Commands(cmds ...Commandor) Option {
+	return func(o *Options) {
+		o.commands = make([]Command, 0, len(cmds))
```
Comment (im-kulikov) [inline cli/options.go]: every time rewrite?

Comment (jenchik) [reply cli/options.go]: I think so, it's better

**Comment by pcherednichenko** [comment]
посмотрел, в целом все ок
поскольку *WIP* смотрел чисто в общем

File: cli/cli.go [hunk 3]
```diff
@@ -26,70 +26,114 @@ var (
...
+		if options.ConfigInterface != nil {
+			if redisConf, ok := hasRedis(options.ConfigInterface); ok {
+				if options.Redis, err = redisConf.Connect(); err != nil {
+					return err
+				}
+			}
+		}
+
+		// Validate options:
+		if err = validator.New().Struct(options); err != nil {
```
Comment (im-kulikov) [inline cli/cli.go]: `return validator.New().Struct(options)`

File: cli/cli.go [hunk 4]
```diff
@@ -26,70 +26,114 @@ var (
...
+	if !ok {
+		// TODO or return an error?
+		return nil
+	}
+
+	if len(dbname) != 0 {
+		dbConf.Database = dbname
+	}
+
+	if opts.DB, err = dbConf.Connect(); err != nil {
```
Comment (im-kulikov) [inline cli/cli.go]:
```go
/...

opts.DB, err = dbConf.Connect()
return err
```

**Review by im-kulikov — CHANGES_REQUESTED** [review]
```sh
→ ./builder init --dir ../something2
Project mame: something2
Project full name: thub.com/cryptopay-dev/something2
Project path: /Users/kulikov/.golang/src/github.com/cryptopay-dev/something2
panic: exit status 1

goroutine 1 [running]:
main.main()
        /Users/kulikov/.golang/src/github.com/cryptopay-dev/yaga/cmd/main.go:100 +0x2ca

```

File: cmd/project_example/app/app.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+type authors struct {
+	Name  string
+	Email string
+}
+
+// App instance
+type App struct {
+	cliApp.RunOptions
+	Config       config.Config
+	Debug        bool
```
Comment (im-kulikov) [inline cmd/project_example/app/app.go]: remove:
- Debug
- BuildTime
- BuildVersion

there is inside cliApp.RunOptions

**Review by im-kulikov — DISMISSED** [review]
I do not quite agree with the use of
- `vars`
- `cp`

but as a first iteration.. I guess it's ok

File: cmd/project_example/app/library/config/config.go [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package config
+
+// Config structure
+type Config struct {
+	Bind string `yaml:"bind" validate:"required"`
```
Comment (im-kulikov) [inline cmd/project_example/app/library/config/config.go]: pls add:

```go
// import config "gihtub.com/yaga/config"
// Database    config.Database `yaml:"database" validate:"required,dive"`
// Redis    config.Redis `yaml:"redis" validate:"required,dive"`
```

File: cmd/project_example/config.example.yml [hunk 7]
```diff
@@ -0,0 +1 @@
+bind: :8080
```
Comment (im-kulikov) [inline cmd/project_example/config.example.yml:1]: pls add:

```yaml
# example database:
#database:
#  address: localhost:5432
#  database: my-db
#  user: postgres
#  password: postgres

# example redis:
#redis:
#  address: 127.0.0.1:6379
#  database: 0
#  password:
#  pool_size: 10
#  pool_timeout: 3
```

**Comment by im-kulikov** [comment]
XD `vars/vars.go` + `misc/vars.go`... more... add more!

**Review by pcherednichenko — DISMISSED** [review]

File: cmd/project_example/Gopkg.toml [hunk 8]
```diff
@@ -0,0 +1,12 @@
+[[constraint]]
+  name = "github.com/cryptopay-dev/yaga"
+  version = "1.8"
+
+[[constraint]]
+  name = "gopkg.in/go-playground/validator.v9"
```
Comment (im-kulikov) [inline cmd/project_example/Gopkg.toml]: for what?... it is already in the Yaga :)

Comment (im-kulikov) [reply cmd/project_example/Gopkg.toml]: https://github.com/cryptopay-dev/yaga/blob/bce5c14b703b19912163d4335d6cd571be9834bf/Gopkg.lock#L238-L242

**Review by im-kulikov — APPROVED** [review]
giphy-0

**Review by pcherednichenko — APPROVED** [review]

**Pull request merged**